core/test/utils/matrix_utils.hpp

```diff
@@ -47,108 +47,221 @@ namespace test {
+ * Replace A by (A + op(A^T)) / 2 in the matrix data
  *
- * @tparam ValueType  valuetype of Dense matrix to process
+ * @param data  the matrix data
+ * @param op  the operator to apply to the transposed value
  *
- * @param mtx  the dense matrix
- * @param ratio  the scale to set the diagonal value. default is 1 and it must
- *               be larger than or equal to 1.
+ * @tparam ValueType  the value type
+ * @tparam IndexType  the index type
  */
-template <typename ValueType>
-void make_diag_dominant(matrix::Dense<ValueType>* mtx,
+template <typename ValueType, typename IndexType, typename Op>
+void make_symmetric_generic(matrix_data<ValueType, IndexType>& data, Op op)
+{
+    GKO_ASSERT_IS_SQUARE_MATRIX(data.size);
+    // compute (A + op(A^T)) / 2
+    const auto nnz = data.nonzeros.size();
+    for (std::size_t i = 0; i < nnz; i++) {
+        data.nonzeros[i].value /= 2.0;
+        auto entry = data.nonzeros[i];
+        data.nonzeros.emplace_back(entry.column, entry.row, op(entry.value));
+    }
+    // combine duplicates
+    data.sum_duplicates();
+}
+
+
+/**
+ * Replace A by (A + A^T) / 2 in the matrix data
+ *
+ * @param data  the matrix data
+ *
+ * @tparam ValueType  the value type
+ * @tparam IndexType  the index type
+ */
+template <typename ValueType, typename IndexType>
+void make_symmetric(matrix_data<ValueType, IndexType>& data)
+{
+    make_symmetric_generic(data, [](auto val) { return val; });
+}
+
+
+/**
+ * Replace A by (A + conj(A^T)) / 2 in the matrix data
+ *
+ * @param data  the matrix data
+ *
+ * @tparam ValueType  the value type
+ * @tparam IndexType  the index type
+ */
+template <typename ValueType, typename IndexType>
+void make_hermitian(matrix_data<ValueType, IndexType>& data)
+{
+    make_symmetric_generic(data, [](auto val) { return gko::conj(val); });
+}
+
+
+/**
+ * Scales the matrix data such that its diagonal is at least ratio times as
+ * large as the sum of offdiagonal magnitudes. Inserts diagonals if missing.
+ *
+ * @param data  the matrix data
+ * @param ratio  how much bigger should the diagonal entries be compared to the
+ *               offdiagonal entry magnitude sum?
+ *
+ * @tparam ValueType  the value type
+ * @tparam IndexType  the index type
+ */
+template <typename ValueType, typename IndexType>
+void make_diag_dominant(matrix_data<ValueType, IndexType>& data,
                         remove_complex<ValueType> ratio = 1.0)
 {
-    // To keep the diag dominant, the ratio should be larger than or equal to 1
     GKO_ASSERT_EQ(ratio >= 1.0, true);
-    auto mtx_host =
-        make_temporary_clone(mtx->get_executor()->get_master(), mtx);
-
-    using std::abs;
-    for (size_type i = 0; i < mtx_host->get_size()[0]; ++i) {
-        auto sum = gko::zero<ValueType>();
-        for (size_type j = 0; j < mtx_host->get_size()[1]; ++j) {
-            sum += abs(mtx_host->at(i, j));
+    GKO_ASSERT_IS_SQUARE_MATRIX(data.size);
+    std::vector<remove_complex<ValueType>> norms(data.size[0]);
```

Comment:
What happens where there is an empty row? Does anything specific need to be done (adding a diagonal entry?) or is keeping the row empty fine?

Reply:
good point. I guess in that case, we should just set the diagonal to something non-zero to make the matrix non-singular